risc0/core/src/inverse.rs

```diff
@@ -0,0 +1,143 @@
+// Copyright 2024 RISC Zero, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use alloc::{vec, vec::Vec};
+use core::cmp::min;
+
+use crate::field::Elem;
+
+struct ExtensionField<T> {
+    degree: usize,
+    irreducible: Vec<T>,
+}
+
+impl<T: Elem> ExtensionField<T> {
+    fn new(degree: usize, irreducible: Vec<T>) -> Self {
+        Self {
+            degree,
+            irreducible,
+        }
+    }
+
+    fn zero(&self) -> Vec<T> {
+        vec![T::ZERO; self.degree]
+    }
+
+    fn one(&self) -> Vec<T> {
+        let mut coeffs = vec![T::ZERO; self.degree];
+        coeffs[0] = T::ONE;
+        coeffs
+    }
+
+    fn sub(&self, a: &[T], b: &[T]) -> Vec<T> {
+        let mut c = a.to_vec();
+        c.resize(self.degree, T::ZERO);
+        for i in 0..min(self.degree, b.len()) {
+            c[i] -= b[i];
+        }
+        c
+    }
+
+    fn mul(&self, a: &[T], b: &[T]) -> Vec<T> {
+        let mut c = self.naive_mul(a, b);
+
+        // Reduce the degree using the irreducible polynomial
+        let upper = 2 * self.degree - 2;
+        for i in (self.degree..=upper).rev() {
+            let x = c[i];
+            for j in 0..self.degree {
+                c[i - self.degree] += x * self.irreducible[j];
+            }
+            c[i] = T::ZERO;
+        }
+        c.truncate(self.degree);
+        c
+    }
+
+    fn mul_elem(&self, a: T, b: &[T]) -> Vec<T> {
+        let mut c = self.zero();
+        for i in 0..self.degree {
+            c[i] = a * b[i];
+        }
+        c
+    }
+
+    fn inv(&self, a: &[T]) -> Vec<T> {
+        let f = ExtensionField::new(self.degree + 1, self.irreducible.clone());
+        let mut t0 = f.zero();
+        let mut t1 = f.one();
+        let mut r0 = self.irreducible.clone();
```

Comment:
There's a small discrepancy here: the C++ inv implementation in the compiler uses `r0 = [p - 11, 0, 0, 0, p - 1]`. I don't remember why that ended up that way, but I do recall testing intermediate states of that algorithm implementation against at least one non-trivial example I worked out by hand so I'm sure there was a good reason.